Paper: Computational Resource Optimisation in Feature Selection under Class Imbalance Conditions

AmadiGabriel opened the pull request:
If you are creating this PR in order to submit a draft of your paper, please name your PR with `Paper: <title>`. An editor will then add a `paper` label and GitHub Actions will be run to check and build your paper.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#preview-your-paper) for more information.

_Editor_: <!--editor-->Chris Calloway @cbcunc <!--end-editor-->

_Reviewers_: <!--reviewers-list-->
- Jane Adams @janeadams
- Andrei Paleyes @apaleyes
<!--end-reviewers-list-->

AmadiGabriel committed:
all changes

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Create images

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
remove images

AmadiGabriel committed:
one image

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
resultspics

AmadiGabriel committed:
aMerge branch '2024' of https://github.com/AmadiGabriel/scipy_proceedings into 2024

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Credit Notes myst.yml

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
removed googlesholarID

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Updated subfigures

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
updated all subfigures

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
check figure rendering

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
included datasets

AmadiGabriel committed:
included classifiers previously omitted

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
time_plot

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
time_plot

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
updated figure numbering_sorting reference display

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update mybib.bib

ejm714 commented:
@AmadiGabriel could you please be mindful in which github action runs you cancel? Looks like you've cancelled two of mine. I'm just trying to get my paper preview updated with a typo fix

<img width="1152" alt="image" src="https://github.com/scipy-conference/scipy_proceedings/assets/22667367/7b69b562-2985-43fe-8215-7a311a010c07">

apaleyes commented:
Hello @AmadiGabriel ! I am one of the reviewers of your paper. Looking forward to reading it, please watch out for some comments here.

Also, a question to chairs (@cbcunc I guess). This PR is coming with lots of data and code in addition to the paper itself. Shall we be reviewing all of that material, or only the manuscript itself?

AmadiGabriel commented:
> Hello @AmadiGabriel ! I am one of the reviewers of your paper. Looking forward to reading it, please watch out for some comments here.
> 
> Also, a question to chairs (@cbcunc I guess). This PR is coming with lots of data and code in addition to the paper itself. Shall we be reviewing all of that material, or only the manuscript itself?

@apaleyes good to know. we would be looking forward to your comments.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+# Ensure that this title is the same as the one in `myst.yml`
+title: Computational Resource Optimisation in Feature Selection under Class Imbalance Conditions
+abstract: |
+  Feature selection is crucial for reducing data dimensionality as well as enhancing model interpretability and performance in machine learning tasks. However, selecting the most informative features in large dataset, often incurs high computational costs. This study explores the possibility of performing feature selection on a subset of data to reduce the computational burden. The study uses fiv ...
```
```suggestion
  Feature selection is crucial for reducing data dimensionality as well as enhancing model interpretability and performance in machine learning tasks. However, selecting the most informative features in large dataset often incurs high computational costs. This study explores the possibility of performing feature selection on a subset of data to reduce the computational burden. The study uses five real-life datasets with substantial sample sizes and severe class imbalance ratios between 0.09 – 0.18. The results illustrate the stability/variability of feature importance with smaller sample fractions in different models. In this study, LightGBM exhibited the most stable performance, even with reduced sample fractions.
```

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+# Ensure that this title is the same as the one in `myst.yml`
+title: Computational Resource Optimisation in Feature Selection under Class Imbalance Conditions
+abstract: |
+  Feature selection is crucial for reducing data dimensionality as well as enhancing model interpretability and performance in machine learning tasks. However, selecting the most informative features in large dataset, often incurs high computational costs. This study explores the possibility of performing feature selection on a subset of data to reduce the computational burden. The study uses fiv ...
```
"large dataset, often incurs" - comma isn't needed here, I believe

AmadiGabriel replied:
Comma removed

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+In the development of prediction models for real-world applications, two key challenges often arise: high-dimensionality resulting from the numerous features, and class-imbalance due to the rarity of samples in the positive class. Feature selection methods are  utilised to address issues of high-dimensionality by selecting a smaller subset of relevant features, thus reducing noise, increasing int ...
+
+Studies [@Matharaarachchi2021; @Tsai2020; @Yin2013; @deHaro-Garcia2020] on the performance of feature selections methods with class imbalance data have been undertaken on using synthetic and real-life datasets. A significant drawback noted was the computational cost of their approach on large sample sizes.  While experimental investigations of feature selection amid class imbalance conditions hav ...
```
"feature selections methods" -> "feature selection methods"

"in the literatures" -> "in the literature"

AmadiGabriel replied:
corrected.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+Studies [@Matharaarachchi2021; @Tsai2020; @Yin2013; @deHaro-Garcia2020] on the performance of feature selections methods with class imbalance data have been undertaken on using synthetic and real-life datasets. A significant drawback noted was the computational cost of their approach on large sample sizes.  While experimental investigations of feature selection amid class imbalance conditions hav ...
+
+This study investigates the impact of performing feature selection on a reduced dataset on feature importance and model performance, using five real-life datasets characterised by large sample sizes and severe class imbalance structures. We employ a rigorous feature selection process that utilises permutation feature importance (PFI) and evaluate the feature importance on LightGBM, random forest  ...
```
First time they are mentioned, LightGBM and SVM probably need to be spelled out fully.

AmadiGabriel replied:
Spelt out and abbreviated appropriately.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+This study investigates the impact of performing feature selection on a reduced dataset on feature importance and model performance, using five real-life datasets characterised by large sample sizes and severe class imbalance structures. We employ a rigorous feature selection process that utilises permutation feature importance (PFI) and evaluate the feature importance on LightGBM, random forest  ...
+
+The rest of the paper is organised as follows: @sec:methodology briefly outlines the methodology adopted and present the results in @sec:results presents the results and discussion. The conclusion of the study is provided in @sec:conclusion.
```
"present the results in" should be deleted

AmadiGabriel replied:
deleted.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+(sec:methodology)=
+## Methodology
+### Description of datasets
+Five real life datasets from different subject areas were considered in this study. Four of the datasets were obtained from the UC Irvine machine learning repository, including CDC Diabetes Health Indicator [@diabetes], Census Income [@census_income], Bank Marketing [@bank_marketing], Statlog (Shuttle) [@statlog]. The fifth dataset is Moisture Absorbed Composite [@Osa-uwagboe2024] from a damage m ...
```
In the previous section "real-life" with a hyphen was used. I'd suggest being consistent across the whole manuscript.

AmadiGabriel replied:
hyphen included consistently

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+(sec:methodology)=
+## Methodology
+### Description of datasets
+Five real life datasets from different subject areas were considered in this study. Four of the datasets were obtained from the UC Irvine machine learning repository, including CDC Diabetes Health Indicator [@diabetes], Census Income [@census_income], Bank Marketing [@bank_marketing], Statlog (Shuttle) [@statlog]. The fifth dataset is Moisture Absorbed Composite [@Osa-uwagboe2024] from a damage m ...
```
"high class imbalance ratios from 0.09 to 0.18" - it's probably obvious, but I think it would not hurt to explain what this says. E.g. class imbalance ratio of 0.09 means that 0.09*N labels are positive and the rest are negative, right?

AmadiGabriel replied:
An explanation of class imbalance ratio included

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+(sec:methodology)=
+## Methodology
+### Description of datasets
+Five real life datasets from different subject areas were considered in this study. Four of the datasets were obtained from the UC Irvine machine learning repository, including CDC Diabetes Health Indicator [@diabetes], Census Income [@census_income], Bank Marketing [@bank_marketing], Statlog (Shuttle) [@statlog]. The fifth dataset is Moisture Absorbed Composite [@Osa-uwagboe2024] from a damage m ...
```
The number of features in these datasets ranges from 7 to 20. One could argue that 7 isn't all that many, and maybe 20 also isn't enough to justify feature selection. It would be great to have a stronger motivation here as to why you have selected these datasets for the study.

AmadiGabriel replied:
We took the datasets from varying real-world subject areas. This has been justified in the text.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+To maintain a model-agnostic approach that is not confined to any specific ML algorithm, this study employed PFI for feature selection. PFI assesses how each feature affects the model's performance by randomly shuffling the values of a feature and noting the resulting decrease in performance. This technique interrupts the link between a feature and its predicted outcome, thus enabling us determin ...
+
+(sec:results)=
+## Results and Discussions
```
I would argue that everything from here until the `process_feature` snippet still belongs to the methodology section, as it basically describes how you performed the analysis.

AmadiGabriel replied:
Highlighted portion has been moved to methodology section.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+(sec:results)=
+## Results and Discussions
+Datasets were loaded using pandas, and categorical features were encoded appropriately. The Spearman correlation matrix was computed and then converted into a distance matrix. Hierarchical clustering was subsequently performed using Ward’s linkage method, and  a threshold for grouping features into clusters was determined through visual inspection of the dendrograms, allowing for the selection of ...
```
"were encoded appropriately" - let's be concrete, e.g. "using one-hot encoding"

"spearman’s ranking" - > capital S

AmadiGabriel replied:
corrected

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+Feature relationship for moisture absorbed composite dataset; (a) hierarchical cluster, Spearman correlation ranking.
+:::
+
+Subsequently, the investigation proceeded in two steps. In step 1,  the entire samples of the respective dataset was used. The dataset was split into training and test sets based on a test-size of 0.25. The respective classifiers were initialised using their default hyper-parameter settings and fitted on the training data. Thereafter, PFI was computed on the fitted model with number of times a fe ...
```
"the entire samples" -> "all samples" or "the entire dataset"

"importance scores" repeated twice

AmadiGabriel replied:
corrected

apaleyes commented on papers/amadi_udu/main.md:
````diff
@@ -0,0 +1,301 @@
+
+```python
+# Define the function for parallel execution
+def process_feature(f_no, selected_features, df):
````
Even though this is a scientific Python conference proceedings, naturally prompting to showcase some code, I am not sure how valuable this snippet is. The only non-trivial thing here would be `permutation_importance` call, everything else is standard sklearn and pandas operations.

If you are looking to highlight some non-trivial part of the project in the paper, perhaps parallelisation config mentioned in the paragraph above? Or the data analysis behind Figure 1 (a,b)?

AmadiGabriel replied:
Code has been adjusted to retain permuation_importance call and parallelisation config. Sklearn and pandas operations have been expunged.

apaleyes commented on papers/amadi_udu/main.md:
````diff
@@ -0,0 +1,301 @@
+```
+
+
+As observed in [Figure 1a](#hiercorr-a), Frequency Centroid and Peak Frequency are in the same cluster with a highly correlated value of 0.957 shown in [Figure 1b](#hiercorr-b) Similarly, Rise Time and Initiation Frequency are clustered with a highly negative correlation of -0.862. Amplitude and Absolute Energy also exhibited a high positive correlation of 0.981. 
````
Missing full stop after "shown in [Figure 1b](#hiercorr-b)"

AmadiGabriel replied:
corrected

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+  </tr>
+  
+  <tr>
+  <th colspan="11">Moisture Absorbed Composites</th>
```
This one does not seem to render correctly

AmadiGabriel replied:
html table tag corrected.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+As observed in [Figure 1a](#hiercorr-a), Frequency Centroid and Peak Frequency are in the same cluster with a highly correlated value of 0.957 shown in [Figure 1b](#hiercorr-b) Similarly, Rise Time and Initiation Frequency are clustered with a highly negative correlation of -0.862. Amplitude and Absolute Energy also exhibited a high positive correlation of 0.981. 
+
+:::{table} Median and IQR importance scores of features.
```
It is difficult to understand what is this table showing and why is that important. Consider expanding the caption, perhaps with more details about analysis (even if duplicating some text). Also at least mention what text is highlighted in bold, and maybe main conclusions to draw from this table.

AmadiGabriel replied:
Meaning for texts highlighted in bold has been included. More clarity has been provided in the cross-reference and table caption.  The table places some numerical figures to the  subsequent figures presented and discussed.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+However, in Bank Marketing dataset, LightGBM and RF identified Feature 1 as a relatively important feature, while SVM considered it insignificant. The mutability of importance scores for the classifiers considered underscores the need to explore multiple classifiers when undertaking a comprehensive investigation of feature importance for feature selection purposes. 
+
+[Figure 2](#db_time-a) shows the PFI process time and corresponding sample fractions for the Diabetes dataset, which has a substantial sample size of 253,680 instances. The results are based on one independent run, with PFI set at 30 feature-permuted repeats. For LightGBM and RF, the PFI process time increased linearly with larger sample fractions, whereas SVM experienced an exponential growth. L ...
```
Please correct me if that's wrong, but I don't think PFI processing time is actually an important metric to look at. Feature selection is done so that training is eventually faster, so performance-wise what would be good to see is increased training speed vs drop in quality (highlighted in many graphs below). Please let me know if I am misinterpreting things here

AmadiGabriel replied:
PFI was computed on the respective fitted estimators, which gives insight to training time. Thus, an increased PFI processing time implies an increased training speed.

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+
+For LightGBM model in [Figure 3a](#ci_boxplot-a), the median decrease in AUC was close to  zero,  indicating that Final Weight had minimal impact on model performance, as noted in @tbl:result_table. Similar results were recorded in [Figure 4a](#bm_boxplot-a) - [c](#bm_boxplot-c) for the Day of Week feature of Bank Marketing dataset, where all models exhibited similarly high feature importance sco ...
+
+PFI can provide insights into the importance of features, but it is susceptible to variability, especially with smaller sample sizes. Thus, complementary feature selection methods, could be explored to validate feature importance. Future work could investigate the variability of features under particular models and sample sizes, with a view to evolving methods of providing a more stable informati ...
```
"Thus, complementary feature selection methods, could be explored to validate feature importance."  - i don't think this sentence needs commas, especially not the second one

AmadiGabriel replied:
corrected

apaleyes commented on papers/amadi_udu/main.md:
```diff
@@ -0,0 +1,301 @@
+PFI process time and corresponding sample fractions for the Diabetes dataset.
+:::
+
+[Figure 3a](#ci_boxplot-a) - [c](#ci_boxplot-c) present the PFI for Final Weight feature of Census Income dataset, evaluated across different sample fractions using LightGBM, RF, and  SVM models, respectively. The decrease in AUC indicates the impact on model performance when Final Weight feature is permuted.  Generally, for smaller sample fractions, there was a higher variability in AUC and prom ...
```
"It is noteworthy that reduced fractions also incur sample rarity effects of class imbalance data" - that observation seems key to me, since class imbalance was used as one of main motivations behind this work. It seems to undermine the dataset-agnostic feature selection process a bit.

AmadiGabriel replied:
This has been expunged.

apaleyes reviewed (commented):
Very nicely written paper, the final render (which I was able to fish out of the build log) is beautiful and easy to read.

Many minor comments are below. I also have a general question for the authors. I struggle a bit to understand the main contribution or take away from this paper. Let's consider some ideas:
* Is it the method itself? The method appears to be a combination of several well-established techniques, without too much customisation.
* Is it the Python implementation of the method? If so, the emphasis should probably be more on the implementation detail, software design and reuse.
* Is it the analysis of feature selection effects on datasets with class imbalance? If so, the paper needs more unified conclusions, general observations that span beyond single dataset. It probably also needs to analyse larger datasets.
* Is it the analysis of the three selected models? If so, the selection of the models either needs to be expanded, or needs a solid justification (e.g. we used these models because they are most popular in a field X).

I think the paper can be improved a lot if it had this single (or multiple) contribution(s) strongly emphasised, motivated, and supported. Very happy to discuss this question more here, in this PR, as a part of the review process.

AmadiGabriel commented:
> Very nicely written paper, the final render (which I was able to fish out of the build log) is beautiful and easy to read.
> 
> Many minor comments are below. I also have a general question for the authors. I struggle a bit to understand the main contribution or take away from this paper. Let's consider some ideas:
> 
> * Is it the method itself? The method appears to be a combination of several well-established techniques, without too much customisation.
> * Is it the Python implementation of the method? If so, the emphasis should probably be more on the implementation detail, software design and reuse.
> * Is it the analysis of feature selection effects on datasets with class imbalance? If so, the paper needs more unified conclusions, general observations that span beyond single dataset. It probably also needs to analyse larger datasets.
> * Is it the analysis of the three selected models? If so, the selection of the models either needs to be expanded, or needs a solid justification (e.g. we used these models because they are most popular in a field X).
> 
> I think the paper can be improved a lot if it had this single (or multiple) contribution(s) strongly emphasised, motivated, and supported. Very happy to discuss this question more here, in this PR, as a part of the review process.

Thank you @apaleyes for these commendation, comments and observations raised. I would say that they are quite thorough and valuable. We would attempt to address all points raised and make clarifications where possible. I have drawn the attention of the co-authors appropriately. Do stand by for our responses within the week.

AmadiGabriel commented:
@rowanc1 could you help check why check is failing?

AmadiGabriel committed:
Update main.md

rowanc1 commented:
@AmadiGabriel, we are working through a change on the action. @fwkoch will be fixing this shortly and we will rerun it for you!

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update myst.yml

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update mybib.bib

AmadiGabriel committed:
table issues

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update pfi.py

AmadiGabriel committed:
Update pfi.py

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update pfi.py

AmadiGabriel committed:
Update pfi.py

AmadiGabriel committed:
Update box_plots.py

AmadiGabriel committed:
Update pfi.py

AmadiGabriel committed:
Update main.md

cbcunc commented:
Review reminder sent to @janeadams

AmadiGabriel committed:
Update main.md

janeadams commented:
A succinct and interesting read on evaluating permutation feature importance (PFI) impacts on three different classification models (Random Forest, LightGBM, and SVM) with varying proportions of subsampled data featuring unbalanced classes. I have minor comments but overall I think this a great contribution.

- The dual axes in the processing time figure were odd to me at first; it might be valuable to explain that SVM's poor performance relative to the other two methods is likely due to its poor parallelizability (if that's a word)
- The "decrease in AUC" figures are confusing in that negative x-axis values must therefore indicate increased in AUC? (Correct me if I am misunderstanding). This forces the reader to think about a "double negative makes a positive" which adds possibly unnecessary complexity to interpretation. I would recommend either 1) changing the axis / measure to just be "change in AUC" and/or 2) adding annotations directly onto the white space with an arrow indicating "poorer performance this direction" or similar.

I particularly appreciated the pre-filtering step of using hierarchical clustering of features to account for potential collinearities. I also appreciated that the authors used multiple data sets and evaluated at a range of sample proportions. This is a nice example of how a lot of scientific computing python libraries can come together into a single interesting experiment.

AmadiGabriel commented:
> A succinct and interesting read on evaluating permutation feature importance (PFI) impacts on three different classification models (Random Forest, LightGBM, and SVM) with varying proportions of subsampled data featuring unbalanced classes. I have minor comments but overall I think this a great contribution.
> 
> * The dual axes in the processing time figure were odd to me at first; it might be valuable to explain that SVM's poor performance relative to the other two methods is likely due to its poor parallelizability (if that's a word)
> * The "decrease in AUC" figures are confusing in that negative x-axis values must therefore indicate increased in AUC? (Correct me if I am misunderstanding). This forces the reader to think about a "double negative makes a positive" which adds possibly unnecessary complexity to interpretation. I would recommend either 1) changing the axis / measure to just be "change in AUC" and/or 2) adding annotations directly onto the white space with an arrow indicating "poorer performance this direction" or similar.
> 
> I particularly appreciated the pre-filtering step of using hierarchical clustering of features to account for potential collinearities. I also appreciated that the authors used multiple data sets and evaluated at a range of sample proportions. This is a nice example of how a lot of scientific computing python libraries can come together into a single interesting experiment.

Thank you for the encouraging comments and observations on the paper @janeadams . We are currentlly addressing some of the comments raised by @apaleyes . Hopefully, all observations raised will be responded to early next week and the paper updated accordingly.

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
merged fig2

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Delete papers/amadi_udu/images/gsvs_cmap.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/gsvs_hierclus.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/gsvs_hierclus_cmap.png

AmadiGabriel committed:
merged fig 2

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Delete papers/amadi_udu/images/rf_statlog_shuttle_feature_0.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/lgbm_statlog_shuttle_feature_0.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/svm_statlog_shuttle_feature_0.png

AmadiGabriel committed:
updated fig5

AmadiGabriel committed:
Delete papers/amadi_udu/images/rf_census_income_feature_2.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/lgbm_census_income_feature_2.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/svm_census_income_feature_2.png

AmadiGabriel committed:
census_income_fig

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Delete papers/amadi_udu/images/lgbm_bank_marketing_feature_5.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/rf_bank_marketing_feature_5.png

AmadiGabriel committed:
Delete papers/amadi_udu/images/svm_bank_marketing_feature_5.png

AmadiGabriel committed:
fig 5 updated

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

AmadiGabriel commented:
> Very nicely written paper, the final render (which I was able to fish out of the build log) is beautiful and easy to read.
> 
> Many minor comments are below. I also have a general question for the authors. I struggle a bit to understand the main contribution or take away from this paper. Let's consider some ideas:
> 
> * Is it the method itself? The method appears to be a combination of several well-established techniques, without too much customisation.
> * Is it the Python implementation of the method? If so, the emphasis should probably be more on the implementation detail, software design and reuse.
> * Is it the analysis of feature selection effects on datasets with class imbalance? If so, the paper needs more unified conclusions, general observations that span beyond single dataset. It probably also needs to analyse larger datasets.
> * Is it the analysis of the three selected models? If so, the selection of the models either needs to be expanded, or needs a solid justification (e.g. we used these models because they are most popular in a field X).
> 
> I think the paper can be improved a lot if it had this single (or multiple) contribution(s) strongly emphasised, motivated, and supported. Very happy to discuss this question more here, in this PR, as a part of the review process.

Thank you @apaleyes for your insightful comments that has enhance the content and quality of the paper. 
Our ideas are relate closely to the latter two questions asked.

We have provided a more unified conclusion that clarifies that the paper is a preliminary study that considers five datasets with substantial sample size, characterised by class imbalance. A justification for the selection of the models has been included in the text. This informed the choice of PFI for the feature selection process owing to its advantage of being model-agnostic. Expansion to include other models and much larger datasets has been included in the conclusion for further study.

> Very nicely written paper, the final render (which I was able to fish out of the build log) is beautiful and easy to read.
> 
> Many minor comments are below. I also have a general question for the authors. I struggle a bit to understand the main contribution or take away from this paper. Let's consider some ideas:
> 
> * Is it the method itself? The method appears to be a combination of several well-established techniques, without too much customisation.
> * Is it the Python implementation of the method? If so, the emphasis should probably be more on the implementation detail, software design and reuse.
> * Is it the analysis of feature selection effects on datasets with class imbalance? If so, the paper needs more unified conclusions, general observations that span beyond single dataset. It probably also needs to analyse larger datasets.
> * Is it the analysis of the three selected models? If so, the selection of the models either needs to be expanded, or needs a solid justification (e.g. we used these models because they are most popular in a field X).
> 
> I think the paper can be improved a lot if it had this single (or multiple) contribution(s) strongly emphasised, motivated, and supported. Very happy to discuss this question more here, in this PR, as a part of the review process.

Thank you @apaleyes for the insightful comments provided to this paper, which has enhanced the quality and richness of the paper.
The main contribution of the work border on the latter two questions you raised. Accordingly, we have provided a more unified conclusion and justified the selection of the models. As this is a preliminary investigation, we have also included as part of future work an expansion to introduce some quantitative measure of the variability of models and feature selection methods. Other comments have also been addressed.

AmadiGabriel committed:
Update main.md

AmadiGabriel commented:
> A succinct and interesting read on evaluating permutation feature importance (PFI) impacts on three different classification models (Random Forest, LightGBM, and SVM) with varying proportions of subsampled data featuring unbalanced classes. I have minor comments but overall I think this a great contribution.
> 
> * The dual axes in the processing time figure were odd to me at first; it might be valuable to explain that SVM's poor performance relative to the other two methods is likely due to its poor parallelizability (if that's a word)
> * The "decrease in AUC" figures are confusing in that negative x-axis values must therefore indicate increased in AUC? (Correct me if I am misunderstanding). This forces the reader to think about a "double negative makes a positive" which adds possibly unnecessary complexity to interpretation. I would recommend either 1) changing the axis / measure to just be "change in AUC" and/or 2) adding annotations directly onto the white space with an arrow indicating "poorer performance this direction" or similar.
> 
> I particularly appreciated the pre-filtering step of using hierarchical clustering of features to account for potential collinearities. I also appreciated that the authors used multiple data sets and evaluated at a range of sample proportions. This is a nice example of how a lot of scientific computing python libraries can come together into a single interesting experiment.

Thank you @janeadams for the observations and review of the paper. This has provided clarity to aspects of the data visualisation and improved the deductions on the model performance. 

An explanation for SVM’s poor performance included in the text.
Axis changed to “change in AUC”. More explanation has been included to clarify positive and negative PFI performance results.

AmadiGabriel committed:
Update main.md

AmadiGabriel committed:
Update main.md

apaleyes commented:
Lovely, thanks for all the work on updating the paper @AmadiGabriel ! I'll have another look shortly

cbcunc commented:
@AmadiGabriel Good to meet you at SciPy. I am inviting you to review this paper. You were sent an invitation from GitHub to be a collaborator on this repository. Please accept the invitation. Your review should be in the form of GitHub review comments: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request

apaleyes commented:
"shortly" ha-ha (one month later)

Anyhow, @cbcunc @ameyxd  I am happy with the changes made to the paper and how the comments were addressed. If I am reading it right some additional experiments were run, quite impressive!

To the authors, I still think the number of features in the datasets reviewed isn't big enough to justify feature selection. It absolutely works as the first step, but would be nice to see the follow up work on larger datasets, as the future work paragraph promises. Same conference next year?

janeadams commented:
Great work! It looks to me like all comments were addressed and the paper has turned out nicely.
